exonum/src/events/noise/mod.rs

```diff
@@ -143,22 +143,29 @@ impl Handshake for NoiseHandshake {
     }
 }
 
-fn read<S: AsyncRead + 'static>(sock: S) -> impl Future<Item = (S, Vec<u8>), Error = io::Error> {
-    let buf = vec![0u8; HANDSHAKE_HEADER_LENGTH];
-    // First byte of handshake message is payload length, remaining bytes [1; len] is
-    // the handshake payload. Therefore, we need to read first byte and after that
-    // remaining payload.
-    read_exact(sock, buf).and_then(|(stream, msg)| read_exact(stream, vec![0u8; msg[0] as usize]))
-}
+pub struct HandshakeRawMessage(Vec<u8>);
+
+impl HandshakeRawMessage {
+    pub fn read<S: AsyncRead + 'static>(
+        sock: S,
+    ) -> impl Future<Item = (S, Self), Error = io::Error> {
+        let buf = vec![0u8; HANDSHAKE_HEADER_LENGTH];
+        // First byte of handshake message is payload length, remaining bytes [1; len] is
+        // the handshake payload. Therefore, we need to read first byte and after that
+        // remaining payload.
+        read_exact(sock, buf)
+            .and_then(|(stream, msg)| read_exact(stream, vec![0u8; msg[0] as usize]))
+            .and_then(|(stream, msg)| Ok((stream, HandshakeRawMessage(msg))))
+    }
 
-fn write<S: AsyncWrite + 'static>(
-    sock: S,
-    buf: &[u8],
-    len: usize,
-) -> impl Future<Item = (S, Vec<u8>), Error = io::Error> {
-    debug_assert!(len < NOISE_MAX_HANDSHAKE_MESSAGE_LENGTH);
+    pub fn write<S: AsyncWrite + 'static>(
+        self,
+        sock: S,
+    ) -> impl Future<Item = (S, Vec<u8>), Error = io::Error> {
+        let len = self.0.len();
+        debug_assert!(len < NOISE_MAX_HANDSHAKE_MESSAGE_LENGTH);
```

Comment:
Nit: I'd rename to `MAX_HANDSHAKE_MESSAGE_LENGTH` (similar to other constants) to avoid stuttering.